presto-main/src/main/java/com/facebook/presto/operator/scalar/KDistinct.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.annotation.UsedByGeneratedCode;
+import com.facebook.presto.common.QualifiedObjectName;
+import com.facebook.presto.metadata.BoundVariables;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.metadata.SqlScalarFunction;
+import com.facebook.presto.spi.function.FunctionKind;
+import com.facebook.presto.spi.function.Signature;
+import com.facebook.presto.spi.function.SqlFunctionVisibility;
+import com.facebook.presto.sql.gen.lambda.BinaryFunctionInterface;
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.metadata.BuiltInTypeAndFunctionNamespaceManager.DEFAULT_NAMESPACE;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.functionTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.valueTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.NullConvention.RETURN_NULL_ON_NULL;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class KDistinct
```

Comment:
Please keep the class name and function name consistent, so maybe change this to `TopkDistinct` instead. Or `TopKDistinct` and `top_k_distinct`, or name the function `k_distinct`.

Reply:
+1

Using the term topk in the name might be confusing as top-k usually refers to k first elements in a sorted list, but here the list is not sorted. I'd also consider adding approx_ prefix to the name to make it crystal clear that results may not be 100% accurate.

Also, I wonder if this logic could be folded into DistinctLimitOperator instead of a function.

Reply:
DistinctLimit operator is quite heavy and it has partitioning semantics etc. So I didn't want to touch it because it is needed for the other case when it is turned off.

And I'm convinced here due to my research :) that for the threshold I have (10K) it's actually accurate. We have a featuresconfig so for deployments that are for adhoc/analytical queries this is fine. It is definitely not approximate.